feg/gateway/sbi/sbi_log_middleware.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2021 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sbi
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/labstack/echo/v4"
+)
+
+// LoggingTransport implements the http.RoundTripper interface and logs the
+// HTTP request and response information
+type LoggingTransport struct {
+	Transport http.RoundTripper
+	Logger    SbiLogger
+}
+
+type responseGrabber struct {
+	io.Writer
+	http.ResponseWriter
+}
+
+func NewLoggingTransport() LoggingTransport {
+	return LoggingTransport{
+		Transport: http.DefaultTransport,
+		Logger:    SbiLogger{},
+	}
+}
+
+func NewLoggingHttpClient() *http.Client {
+	return &http.Client{
+		Transport: NewLoggingTransport(),
+	}
+}
+
+// RoundTrip overrides the http.DefaultTransport RoundTrip behavior and logs the HTTP request and response.
+func (lt LoggingTransport) RoundTrip(req *http.Request) (res *http.Response, err error) {
+	// Before sending
+	var reqBody []byte
+
+	if req.Body != nil {
+		reqBody, err = ioutil.ReadAll(req.Body)
+		if err != nil {
+			return
+		}
+		// reset the req.Body back to original
+		req.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody))
+	}
+	lt.Logger.LogRequest(req.Method, req.URL, reqBody, req.Header)
+
+	// Make the actual request
+	start := time.Now()
+	res, err = lt.Transport.RoundTrip(req)
+	took := time.Since(start)
+	if err != nil {
+		// error responses shall be logged by caller
+		return
+	}
+
+	// After sending
+	var resBody []byte
+	if res.Body != nil {
+		resBody, err = ioutil.ReadAll(res.Body)
+		if err != nil {
+			return
+		}
+		// reset the res.Body back to original
+		res.Body = ioutil.NopCloser(bytes.NewBuffer(resBody))
+	}
+	lt.Logger.LogResponse(req.URL, res.Status, resBody, res.Header, took)
+	return
+}
```

Comment:
I am concerned about this because this code is only used when logging. That means that if for any reason there is a bug there, we will only be able to see it when v==2. 

Instead of checking `glog.V(2)` outside, wouldn't it make sense to check it on the function? 
This way you just need to always add this custom logger on sbi. The logger will take care of printing or not printing.  Plus we will always be using this `RoundTrip` function and `NewLoggingHttpClient` object

Let me know if that doesn't make sense to you. Maybe I am missing something


```
func (lt LoggingTransport) RoundTrip(req *http.Request) (res *http.Response, err error) {
	// Before sending
	var reqBody []byte


	if glog.V(2) {
		  if req.Body != nil {
			  reqBody, err = ioutil.ReadAll(req.Body)
			  if err != nil {
				  return
			  }
			  // reset the req.Body back to original
			  req.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody))
		  }
		  lt.Logger.LogRequest(req.Method, req.URL, reqBody, req.Header)
	}
	// Make the actual request
	start := time.Now()
	res, err = lt.Transport.RoundTrip(req)
	took := time.Since(start)
	if err != nil {
		// error responses shall be logged by caller
		return
	}

	// After sending
	var resBody []byte
	if glog.V(2) { 
		  if res.Body != nil {
			  resBody, err = ioutil.ReadAll(res.Body)
			  if err != nil {
				  return
			  }
			  // reset the res.Body back to original
			  res.Body = ioutil.NopCloser(bytes.NewBuffer(resBody))
		  }
		  lt.Logger.LogResponse(req.URL, res.Status, resBody, res.Header, took)
	}
	return
}
```